context/src/repo/mod.rs

```diff
@@ -0,0 +1,198 @@
+            repo_head_author_email,
+        })
+    }
+
+    fn git_head_branch_from_remote_branches(
+        git_repo: &Repository,
+    ) -> anyhow::Result<Option<String>> {
+        for remote_branch in git_repo
+            .references()?
+            .remote_branches()?
+            .filter_map(Result::ok)
+        {
+            if let Some(target_id) = remote_branch.target().try_id() {
+                if target_id.as_bytes() == remote_branch.id().as_bytes() {
+                    return Ok(remote_branch.name().to_path().to_str().map(String::from));
+                }
+            }
+        }
+        Ok(None)
+    }
+}
+
+#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
+pub struct RepoUrlParts {
+    pub host: String,
+    pub owner: String,
+    pub name: String,
+}
+
+impl RepoUrlParts {
+    pub fn from_url(url: &str) -> anyhow::Result<Self> {
+        let re1 = Regex::new(r"^(ssh|git|http|https|ftp|ftps)://([^/]*?@)?([^/]*)/(.+)/([^/]+)")?;
+        let re2 = Regex::new(r"^([^/]*?@)([^/]*):(.+)/([^/]+)")?;
+
+        let parts = if re1.is_match(url) {
+            let caps = re1.captures(url).expect("failed to parse url");
+            if caps.len() != 6 {
+                return Err(anyhow::anyhow!(
+                    "Invalid repo url format. Expected 6 parts: {:?} (url: {})",
+                    caps,
+                    url
+                ));
+            }
+            let domain = caps.get(3).map(|m| m.as_str()).unwrap_or("");
+            let owner = caps.get(4).map(|m| m.as_str()).unwrap_or("");
+            let name = caps.get(5).map(|m| m.as_str()).unwrap_or("");
+            (domain, owner, name)
+        } else if re2.is_match(url) {
+            let caps = re2.captures(url).expect("failed to parse url");
+            if caps.len() != 5 {
+                return Err(anyhow::anyhow!(
+                    "Invalid repo url format. Expected 6 parts: {:?} (url: {})",
+                    caps,
+                    url
+                ));
+            }
+            let domain = caps.get(2).map(|m| m.as_str()).unwrap_or("");
+            let owner = caps.get(3).map(|m| m.as_str()).unwrap_or("");
+            let name = caps.get(4).map(|m| m.as_str()).unwrap_or("");
+            (domain, owner, name)
+        } else {
+            return Err(anyhow::anyhow!("Invalid repo url format: {}", url));
+        };
+
+        let host = parts.0.trim().to_string();
+        let owner = parts.1.trim().to_string();
+        let name = parts
+            .2
+            .trim()
+            .trim_end_matches('/')
+            .trim_end_matches(".git")
+            .to_string();
+
+        if host.is_empty() || owner.is_empty() || name.is_empty() {
+            return Err(anyhow::anyhow!(
+                "Invalid repo url format. Expected non-empty parts: {:?} (url: {})",
+                parts,
+                url
+            ));
+        }
+
+        Ok(Self { host, owner, name })
```

Comment:
I only copied this logic, but we should probably follow some best practices here, like:
1. Use the [`url`](https://docs.rs/url/latest/url/) crate
2. [Use `lazy_static` for regexes](https://docs.rs/regex/1.11.0/regex/#avoid-re-compiling-regexes-especially-in-a-loop)